File: go/pkg/antfly/lib/clock/clock.go
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

package clock

import (
	"sync"
	"time"
)

// Clock abstracts wall-clock access and timer creation so core control-plane
// loops can be driven deterministically in tests and simulations.
type Clock interface {
	Now() time.Time
	After(d time.Duration) <-chan time.Time
	NewTimer(d time.Duration) Timer
	NewTicker(d time.Duration) Ticker
	Sleep(d time.Duration)
}

// Timer abstracts time.Timer.
type Timer interface {
	C() <-chan time.Time
	Stop() bool
	Reset(d time.Duration) bool
}

// Ticker abstracts time.Ticker.
type Ticker interface {
	C() <-chan time.Time
	Stop()
}

// RealClock delegates to the standard library time package.
type RealClock struct{}

func (RealClock) Now() time.Time {
	return time.Now()
}

func (RealClock) After(d time.Duration) <-chan time.Time {
	return time.After(d)
}

func (RealClock) NewTimer(d time.Duration) Timer {
	return &realTimer{timer: time.NewTimer(d)}
}

func (RealClock) NewTicker(d time.Duration) Ticker {
	return &realTicker{ticker: time.NewTicker(d)}
}

func (RealClock) Sleep(d time.Duration) {
	time.Sleep(d)
}

type realTimer struct {
	timer *time.Timer
}

func (t *realTimer) C() <-chan time.Time {
	return t.timer.C
}

func (t *realTimer) Stop() bool {
	return t.timer.Stop()
}

func (t *realTimer) Reset(d time.Duration) bool {
	return t.timer.Reset(d)
}

type realTicker struct {
	ticker *time.Ticker
}

func (t *realTicker) C() <-chan time.Time {
	return t.ticker.C
}

func (t *realTicker) Stop() {
	t.ticker.Stop()
}

// MockClock advances only when instructed by tests.
type MockClock struct {
	mu      sync.Mutex
	current time.Time
	timers  map[*mockTimer]struct{}
	tickers map[*mockTicker]struct{}
}

func NewMockClock(start time.Time) *MockClock {
	return &MockClock{
		current: start,
		timers:  make(map[*mockTimer]struct{}),
		tickers: make(map[*mockTicker]struct{}),
	}
}

func (m *MockClock) Now() time.Time {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.current
}

func (m *MockClock) After(d time.Duration) <-chan time.Time {
	return m.NewTimer(d).C()
}

func (m *MockClock) NewTimer(d time.Duration) Timer {
	m.mu.Lock()
	defer m.mu.Unlock()

	timer := &mockTimer{
		clock:    m,
		c:        make(chan time.Time, 1),
		deadline: m.current.Add(d),
		active:   true,
	}
	m.timers[timer] = struct{}{}
	return timer
}

func (m *MockClock) NewTicker(d time.Duration) Ticker {
	m.mu.Lock()
	defer m.mu.Unlock()

	ticker := &mockTicker{
		clock:    m,
		c:        make(chan time.Time, 1),
		interval: d,
		next:     m.current.Add(d),
		active:   true,
	}
	m.tickers[ticker] = struct{}{}
	return ticker
}

func (m *MockClock) Sleep(d time.Duration) {
	m.Advance(d)
}

func (m *MockClock) Advance(d time.Duration) {
	m.mu.Lock()
	m.current = m.current.Add(d)
	m.fireDueLocked()
	m.mu.Unlock()
}

func (m *MockClock) Set(t time.Time) {
	m.mu.Lock()
	m.current = t
	m.fireDueLocked()
	m.mu.Unlock()
}

func (m *MockClock) fireDueLocked() {
	for timer := range m.timers {
		if !timer.active || timer.deadline.After(m.current) {
			continue
		}
		timer.active = false
		select {
		case timer.c <- timer.deadline:
		default:
		}
	}

	for ticker := range m.tickers {
		if !ticker.active {
			continue
		}
		for !ticker.next.After(m.current) {
			next := ticker.next
			ticker.next = ticker.next.Add(ticker.interval)
			select {
			case ticker.c <- next:
			default:
			}
		}
	}
}

type mockTimer struct {
	clock    *MockClock
	c        chan time.Time
	deadline time.Time
	active   bool
}

func (t *mockTimer) C() <-chan time.Time {
	return t.c
}

func (t *mockTimer) Stop() bool {
	t.clock.mu.Lock()
	defer t.clock.mu.Unlock()
	wasActive := t.active
	t.active = false
	return wasActive
}

func (t *mockTimer) Reset(d time.Duration) bool {
	t.clock.mu.Lock()
	defer t.clock.mu.Unlock()
	wasActive := t.active
	t.deadline = t.clock.current.Add(d)
	t.active = true
	t.clock.fireDueLocked()
	return wasActive
}

type mockTicker struct {
	clock    *MockClock
	c        chan time.Time
	interval time.Duration
	next     time.Time
	active   bool
}

func (t *mockTicker) C() <-chan time.Time {
	return t.c
}

func (t *mockTicker) Stop() {
	t.clock.mu.Lock()
	defer t.clock.mu.Unlock()
	t.active = false
}

File: zig/lib/platform/src/clock.zig
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

const std = @import("std");
const builtin = @import("builtin");

pub const Clock = struct {
    ctx: ?*anyopaque = null,
    now_realtime_ns_fn: *const fn (?*anyopaque) u64,
    sleep_ms_fn: *const fn (?*anyopaque, u64) void,

    pub fn real() Clock {
        if (builtin.os.tag == .freestanding) {
            return .{
                .now_realtime_ns_fn = freestandingNowRealtimeNs,
                .sleep_ms_fn = freestandingSleepMs,
            };
        }
        return .{
            .now_realtime_ns_fn = realNowRealtimeNs,
            .sleep_ms_fn = realSleepMs,
        };
    }

    pub fn nowRealtimeNs(self: Clock) u64 {
        return self.now_realtime_ns_fn(self.ctx);
    }

    pub fn nowRealtimeMs(self: Clock) u64 {
        return @intCast(self.nowRealtimeNs() / std.time.ns_per_ms);
    }

    pub fn sleepMs(self: Clock, ms: u64) void {
        self.sleep_ms_fn(self.ctx, ms);
    }
};

pub const ManualClock = struct {
    mutex: std.atomic.Mutex = .unlocked,
    now_realtime_ns: u64 = 0,

    pub fn clock(self: *ManualClock) Clock {
        return .{
            .ctx = self,
            .now_realtime_ns_fn = manualNowRealtimeNs,
            .sleep_ms_fn = manualSleepMs,
        };
    }

    pub fn setRealtimeNs(self: *ManualClock, now_realtime_ns: u64) void {
        lockAtomic(&self.mutex);
        defer self.mutex.unlock();
        self.now_realtime_ns = now_realtime_ns;
    }

    pub fn advanceMs(self: *ManualClock, delta_ms: u64) void {
        self.advanceNs(delta_ms * std.time.ns_per_ms);
    }

    pub fn advanceNs(self: *ManualClock, delta_ns: u64) void {
        lockAtomic(&self.mutex);
        defer self.mutex.unlock();
        self.now_realtime_ns +%= delta_ns;
    }
};

fn realNowRealtimeNs(_: ?*anyopaque) u64 {
    var ts: std.posix.timespec = undefined;
    switch (std.posix.errno(std.posix.system.clock_gettime(.REALTIME, &ts))) {
        .SUCCESS => {},
        else => unreachable,
    }
    return @as(u64, @intCast(ts.sec)) * std.time.ns_per_s + @as(u64, @intCast(ts.nsec));
}

fn realSleepMs(_: ?*anyopaque, ms: u64) void {
    var io_impl = std.Io.Threaded.init(std.heap.page_allocator, .{});
    defer io_impl.deinit();
    const sleep_ms = if (ms == 0) @as(u64, 1) else ms;
    std.Io.Clock.Duration.sleep(.{
        .clock = .awake,
        .raw = .fromMilliseconds(@intCast(sleep_ms)),
    }, io_impl.io()) catch {};
}

fn freestandingNowRealtimeNs(_: ?*anyopaque) u64 {
    return 0;
}

fn freestandingSleepMs(_: ?*anyopaque, _: u64) void {}

fn manualNowRealtimeNs(ctx: ?*anyopaque) u64 {
    const self: *ManualClock = @ptrCast(@alignCast(ctx.?));
    lockAtomic(&self.mutex);
    defer self.mutex.unlock();
    return self.now_realtime_ns;
}

fn manualSleepMs(ctx: ?*anyopaque, ms: u64) void {
    _ = ctx;
    _ = ms;
    if (builtin.os.tag != .freestanding) std.Thread.yield() catch {};
}

fn lockAtomic(mutex: *std.atomic.Mutex) void {
    while (!mutex.tryLock()) std.atomic.spinLoopHint();
}
